app/celery/process_pinpoint_receipt_tasks.py

```diff
@@ -42,36 +46,52 @@
 }
 
 
-def event_type_is_optout(event_type, reference):
-    is_optout = event_type == '_SMS.OPTOUT'
-
-    if is_optout:
-        current_app.logger.info(
-            f"event type is OPTOUT for notification with reference {reference})"
-        )
-    return is_optout
-
-
 def _map_record_status_to_notification_status(record_status):
     return _record_status_status_mapping[record_status]
 
 
 @notify_celery.task(bind=True, name="process-pinpoint-result", max_retries=5, default_retry_delay=300)
 @statsd(namespace="tasks")
 def process_pinpoint_results(self, response):
+    """
+    Process a Pinpoint SMS stream event.  Messages long enough to require multiple segments only
+    result in one event that contains the aggregate cost.
+
+    https://docs.aws.amazon.com/pinpoint/latest/developerguide/event-streams-data-sms.html
+    """
+
     if not is_feature_enabled(FeatureFlag.PINPOINT_RECEIPTS_ENABLED):
-        current_app.logger.info('Pinpoint receipts toggle is disabled, skipping callback task')
+        current_app.logger.info('Pinpoint receipts toggle is disabled.  Skipping callback task.')
         return True
 
     try:
         pinpoint_message = json.loads(base64.b64decode(response['Message']))
-        reference = pinpoint_message['attributes']['message_id']
-        event_type = pinpoint_message.get('event_type')
-        record_status = pinpoint_message['attributes']['record_status']
-        current_app.logger.info(
-            f'received callback from Pinpoint with event_type of {event_type} and record_status of {record_status}'
-            f' with reference {reference}'
-        )
+    except (json.decoder.JSONDecodeError, ValueError, TypeError) as e:
+        current_app.logger.exception(e)
+        self.retry(queue=QueueNames.RETRY)
+        return None
+
+    try:
+        pinpoint_attributes = pinpoint_message["attributes"]
+        reference = pinpoint_attributes["message_id"]
+        event_type = pinpoint_message["event_type"]
+        record_status = pinpoint_attributes["record_status"]
+        number_of_message_parts = pinpoint_attributes["number_of_message_parts"]
+        price_in_millicents_usd = pinpoint_message["metrics"]["price_in_millicents_usd"]
+    except KeyError as e:
+        current_app.logger.error("The event stream message data is missing expected attributes.")
+        current_app.logger.exception(e)
+        current_app.logger.debug(pinpoint_message)
+        self.retry(queue=QueueNames.RETRY)
```

Comment:
We discussed this on Zoom.  The retry probably isn't appropriate because the task should always fail if we receive bad data from AWS, but this line is here, in this PR, to preserve the existing behavior.  @k-macmillan will create a ticket to revisit this.